tiledb/sm/serialization/array_schema.cc

```diff
@@ -786,42 +875,26 @@ tuple<Status, optional<shared_ptr<ArraySchema>>> array_schema_deserialize(
         ::capnp::FlatArrayMessageReader reader(kj::arrayPtr(
             reinterpret_cast<const ::capnp::word*>(mBytes),
             serialized_buffer.size() / sizeof(::capnp::word)));
-        capnp::ArraySchema::Reader array_schema_reader =
-            reader.getRoot<capnp::ArraySchema>();
-        RETURN_NOT_OK_TUPLE(
-            array_schema_from_capnp(array_schema_reader, &decoded_array_schema),
-            nullopt);
+        array_schema_reader = reader.getRoot<capnp::ArraySchema>();
         break;
       }
       default: {
-        return {
-            LOG_STATUS(Status_SerializationError(
-                "Error deserializing array schema; Unknown serialization type "
-                "passed")),
-            nullopt};
+        throw StatusException(Status_SerializationError(
+            "Error deserializing array schema; Unknown serialization type "
+            "passed"));
       }
     }
-
-    if (decoded_array_schema == nullptr)
-      return {
-          LOG_STATUS(Status_SerializationError(
-              "Error serializing array schema; deserialized schema is null")),
-          nullopt};
-
-    array_schema = std::move(decoded_array_schema);
   } catch (kj::Exception& e) {
-    return {LOG_STATUS(Status_SerializationError(
-                "Error deserializing array schema; kj::Exception: " +
-                std::string(e.getDescription().cStr()))),
-            nullopt};
+    throw StatusException(Status_SerializationError(
+        "Error deserializing array schema; kj::Exception: " +
+        std::string(e.getDescription().cStr())));
   } catch (std::exception& e) {
-    return {LOG_STATUS(Status_SerializationError(
-                "Error deserializing array schema; exception " +
-                std::string(e.what()))),
-            nullopt};
+    throw StatusException(Status_SerializationError(
+        "Error deserializing array schema; exception " +
+        std::string(e.what())));
   }
 
-  return {Status::Ok(), array_schema};
+  return array_schema_from_capnp(array_schema_reader, URI());
```

Comment:
I believe the problem identified by ASAN is that `message_builder` allocated in the scope of the `try` block above goes out of scope before this call, so the memory referenced by `array_schema_reader` is no longer valid here.